src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,310 @@
+#include "corecel/cont/Array.hh"
+#include "corecel/math/Algorithms.hh"
+#include "orange/OrangeTypes.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+using constants::pi;
+//---------------------------------------------------------------------------//
+/*!
+ * Find positive, real, nonzero roots for involute intersection function.
+ *
+ * The involute intersection equation \f[
+   r_b * [v{cos(t+a)+tsin(t+a)} + u{sin(t+a)-tcos(t+a)}] + xv - yu = 0
+ * \f]
+ * has n solutions mathematically, but we only want solutions where t is
+ results
+ * in a real, positive, and in the defined bounds.  Furthermore the equation is
+ * subject to catastrophic roundoff due to floating point precision (see
+ * \c Tolerance::sqrt_quadratic and the derivation in \c CylAligned ).
+ *
+ * \return An Intersections array where each item is a positive valid
+ * intersection or the sentinel result \c no_intersection() .
+ */
+class InvoluteSolver
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Intersections = Array<real_type, 3>;
+
+  public:
+    // Construct with
+    inline CELER_FUNCTION InvoluteSolver(real_type r_b_,
+                                         real_type a_,
+                                         real_type sign_,
+                                         real_type tmin_,
+                                         real_type tmax_);
+
+    // Solve fully general case
+    inline CELER_FUNCTION Intersections operator()(Real3 const& pos,
+                                                   Real3 const& dir) const;
+
+    //// ACCESSORS ////
+
+    //! Get involute parameters
+    CELER_FUNCTION real_type r_b() const { return r_b_; }
+    CELER_FUNCTION real_type a() const { return a_; }
+    CELER_FUNCTION real_type sign() const { return sign_; }
+
+    //! Get bounds of the involute
+    CELER_FUNCTION real_type tmin() const { return tmin_; }
+    CELER_FUNCTION real_type tmax() const { return tmax_; }
+
+  private:
+    //// DATA ////
+    // Involute parameters
+    real_type r_b_;
+    real_type a_;
+    real_type sign_;
+
+    // Bounds
+    real_type tmin_;
+    real_type tmax_;
+};
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from radius.
+ */
+CELER_FUNCTION InvoluteSolver::InvoluteSolver(
+    real_type r_b, real_type a, real_type sign, real_type tmin, real_type tmax)
+    : r_b_(r_b), a_(a), sign_(sign), tmin_(tmin), tmax_(tmax)
+{
+    CELER_EXPECT(r_b > 0);
+    CELER_EXPECT(a > 0);
+    CELER_EXPECT(std::fabs(tmax) < 2 * pi + std::fabs(tmin));
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Find all positive roots for involute surfaces that are within the bounds.
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(Real3 const& pos,
```

Comment:
This method needs more descriptive comments and explanation of the iteration method

Reply:
I assume you are going to add more comments throughout this function as well so someone can follow the solver procedure?